File: reference/internal/metrics/transitions.go
package metrics

import (
	"sort"
	"strings"
	"time"

	"github.com/Elpulgo/azdo/internal/azdevops"
)

// GapAction tells the snapshot writer what to do for a given prev→curr state
// transition. "Write" = today's observation is enough. "NeedsFallback" =
// intermediate snapshot rows have to be synthesized from /updates because the
// state jumped more than one step or moved backwards.
type GapAction int

const (
	GapWrite         GapAction = iota // normal single-step transition (or first observation)
	GapNeedsFallback                  // multi-step forward jump or backward transition
)

// ClassifyTransition decides whether the writer can append today's observation
// directly, or whether the gap-fallback path has to fire /updates to recover
// the missing intermediate states.
//
// Rules:
//   - first observation (prev empty)              → write
//   - same state                                  → write
//   - legal single forward step in states.Order() → write
//   - everything else (multi-step jump, backward,
//     unknown state)                              → fallback
//
// `states` carries the configured workflow so teams with custom state names
// (e.g. Active / RFT / Done) get the same gap-detection behavior.
func ClassifyTransition(prev, curr string, states StateConfig) GapAction {
	if prev == "" {
		return GapWrite
	}
	if strings.EqualFold(strings.TrimSpace(prev), strings.TrimSpace(curr)) {
		return GapWrite
	}
	pi, pok := states.Index(prev)
	ci, cok := states.Index(curr)
	if !pok || !cok {
		return GapNeedsFallback
	}
	if ci == pi+1 {
		return GapWrite
	}
	return GapNeedsFallback
}

// StateTransition is a single state change at a point in time, extracted from
// the /updates revision history. Used by both the gap-fallback (PR 2) and the
// first-launch backfill (PR 3 — same struct, same fold helper).
type StateTransition struct {
	State string
	At    time.Time
}

// FromAzDevTransitions converts the azdevops layer's transition shape into the
// metrics layer's shape so the pure fold helpers in this package don't have to
// depend on azdevops types at the call site.
func FromAzDevTransitions(in []azdevops.WorkItemStateTransition) []StateTransition {
	out := make([]StateTransition, len(in))
	for i, t := range in {
		out[i] = StateTransition{State: t.State, At: t.At}
	}
	return out
}

// SynthesizeGapRows produces one daily snapshot row per calendar day strictly
// between `since` (exclusive) and `until` (exclusive), filling in the state
// the item was in on each of those days according to `transitions`. The row
// for `until` itself is NOT emitted — the caller already has today's
// observation and will write it separately.
//
// The template carries the static per-item fields (ID, Project, Tags, Points,
// AssignedTo, Iteration); only TS, State, StateSince and Source vary per row.
// All emitted rows are marked Source="updates".
//
// Pure helper — reused verbatim by the PR 3 backfill, which calls it with
// `since` = 90 days ago and `until` = today.
func SynthesizeGapRows(transitions []StateTransition, since, until time.Time, template Snapshot) []Snapshot {
	if len(transitions) == 0 {
		return nil
	}

	sorted := append([]StateTransition(nil), transitions...)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].At.Before(sorted[j].At) })

	// Walk one calendar day at a time across the (since, until) range.
	// For each day, find the latest transition whose `At` is on or before
	// that day, and emit a row stamped with that transition's state.
	startDay := stripTime(since).AddDate(0, 0, 1) // exclusive on `since`
	endDay := stripTime(until)                    // exclusive on `until`

	var rows []Snapshot
	for d := startDay; d.Before(endDay); d = d.AddDate(0, 0, 1) {
		state, stateSince, ok := stateOn(sorted, d)
		if !ok {
			continue
		}
		row := template
		row.TS = d.Format("2006-01-02")
		row.State = state
		row.StateSince = stateSince.UTC().Format(time.RFC3339)
		row.Source = SourceUpdates
		rows = append(rows, row)
	}
	return rows
}

// stateOn returns the state the item was in at the end of day `day`, plus the
// transition timestamp that put it there. Returns ok=false if the item had not
// yet been created on that day (i.e. all transitions are strictly after it).
func stateOn(sorted []StateTransition, day time.Time) (state string, since time.Time, ok bool) {
	// Latest transition whose timestamp is on or before the END of `day`.
	endOfDay := day.Add(24*time.Hour - time.Second)
	for i := len(sorted) - 1; i >= 0; i-- {
		if !sorted[i].At.After(endOfDay) {
			return sorted[i].State, sorted[i].At, true
		}
	}
	return "", time.Time{}, false
}

func stripTime(t time.Time) time.Time {
	return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, t.Location())
}

File: src/Azdo.Core/Metrics/Transitions.cs
using System.Globalization;
using Azdo.Core.AzureDevOps;

namespace Azdo.Core.Metrics;

/// <summary>
/// Tells the snapshot writer what to do for a given prev→curr state transition.
/// </summary>
public enum GapAction
{
    /// <summary>Normal single-step transition (or first observation).</summary>
    Write = 0,

    /// <summary>Multi-step forward jump or backward transition.</summary>
    NeedsFallback,
}

/// <summary>A single state change at a point in time, extracted from /updates history.</summary>
public sealed class StateTransition
{
    public string State { get; set; } = "";
    public DateTime At { get; set; }

    public StateTransition() { }

    public StateTransition(string state, DateTime at)
    {
        State = state;
        At = at;
    }
}

/// <summary>Pure transition classification and gap-row synthesis helpers.</summary>
public static class Transitions
{
    /// <summary>
    /// Decides whether the writer can append today's observation directly, or
    /// whether the gap-fallback path has to fire /updates to recover missing
    /// intermediate states.
    /// </summary>
    public static GapAction ClassifyTransition(string prev, string curr, StateConfig states)
    {
        if (string.IsNullOrEmpty(prev))
            return GapAction.Write;
        if (string.Equals((prev ?? "").Trim(), (curr ?? "").Trim(), StringComparison.OrdinalIgnoreCase))
            return GapAction.Write;
        var (pi, pok) = states.IndexOf(prev!);
        var (ci, cok) = states.IndexOf(curr ?? "");
        if (!pok || !cok)
            return GapAction.NeedsFallback;
        if (ci == pi + 1)
            return GapAction.Write;
        return GapAction.NeedsFallback;
    }

    /// <summary>Converts azdevops transitions into the metrics-layer shape.</summary>
    public static List<StateTransition> FromAzDevTransitions(IReadOnlyList<WorkItemStateTransition> input)
    {
        var outList = new List<StateTransition>(input.Count);
        foreach (var t in input)
            outList.Add(new StateTransition(t.State, t.At));
        return outList;
    }

    /// <summary>
    /// Produces one daily snapshot row per calendar day strictly between
    /// <paramref name="since"/> (exclusive) and <paramref name="until"/> (exclusive),
    /// filling in the state the item was in on each of those days. The row for
    /// <paramref name="until"/> itself is NOT emitted. All emitted rows are marked
    /// Source="updates".
    /// </summary>
    public static List<Snapshot> SynthesizeGapRows(
        IReadOnlyList<StateTransition> transitions, DateTime since, DateTime until, Snapshot template)
    {
        var rows = new List<Snapshot>();
        if (transitions == null || transitions.Count == 0)
            return rows;

        var sorted = transitions.ToList();
        sorted.Sort((a, b) => a.At.CompareTo(b.At));

        var startDay = StripTime(since).AddDays(1); // exclusive on since
        var endDay = StripTime(until);              // exclusive on until

        for (var d = startDay; d < endDay; d = d.AddDays(1))
        {
            if (!StateOn(sorted, d, out var state, out var stateSince))
                continue;
            var row = template.Clone();
            row.TS = d.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
            row.State = state;
            row.StateSince = stateSince.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ssZ", CultureInfo.InvariantCulture);
            row.Source = SourceUpdatesConst;
            rows.Add(row);
        }
        return rows;
    }

    private const string SourceUpdatesConst = Snapshots.SourceUpdates;

    /// <summary>
    /// The state the item was in at the end of day <paramref name="day"/>, plus
    /// the transition timestamp that put it there. ok=false if the item had not
    /// yet been created on that day.
    /// </summary>
    private static bool StateOn(List<StateTransition> sorted, DateTime day, out string state, out DateTime since)
    {
        var endOfDay = day.Add(TimeSpan.FromHours(24) - TimeSpan.FromSeconds(1));
        for (int i = sorted.Count - 1; i >= 0; i--)
        {
            if (sorted[i].At <= endOfDay)
            {
                state = sorted[i].State;
                since = sorted[i].At;
                return true;
            }
        }
        state = "";
        since = default;
        return false;
    }

    private static DateTime StripTime(DateTime t) =>
        new(t.Year, t.Month, t.Day, 0, 0, 0, t.Kind);
}
